pkg/gatewayserver/gatewayserver_test.go

```diff
@@ -556,17 +556,44 @@ func TestGatewayServer(t *testing.T) {
 					},
 				} {
 					t.Run(ctc.Name, func(t *testing.T) {
-						ctx, cancel := context.WithDeadline(ctx, time.Now().Add(timeout))
+						ctx, cancel := context.WithCancel(ctx)
 						upCh := make(chan *ttnpb.GatewayUp)
 						downCh := make(chan *ttnpb.GatewayDown)
-						err := ptc.Link(ctx, t, ctc.ID, ctc.Key, upCh, downCh)
-						cancel()
-						if errors.IsDeadlineExceeded(err) {
-							if !ptc.TimeoutOnInvalidAuth && !ptc.ValidAuth(ctx, ctc.ID, ctc.Key) {
-								t.Fatal("Expected link error due to invalid auth")
+
+						upEvents := map[string]events.Channel{}
+						for _, event := range []string{"gs.gateway.connect"} {
+							upEvents[event] = make(events.Channel, 3)
+						}
+						defer test.SetDefaultEventsPubSub(&test.MockEventPubSub{
+							PublishFunc: func(ev events.Event) {
+								switch name := ev.Name(); name {
+								case "gs.gateway.connect":
+									go func() {
+										upEvents[name] <- ev
+									}()
+								default:
+									t.Logf("%s event published", name)
+								}
+							},
+						})()
+
+						connected := false
+						go func() {
+							select {
+							case <-upEvents["gs.gateway.connect"]:
+								connected = true
```

Comment:
Hmm why not select this channel read after `Link()`, to avoid a dirty write to `connected`?

Reply:
We expect `Link` to return after the context has been cancelled, with in turn must happen after reading the `gs.gateway.connect` event. Dropped the `connected` variable though, and switched to using signal channels for errors.

Decided to use channels there because calling `t.Fatal()` inside goroutines leads to panics when running the tests.